tests/test_main.py

```diff
@@ -30,3 +35,111 @@ def test_set_base64_default_line_size_rejects_negative_values(self):
         with self.assertRaises(ValueError):
             xmlsec.base64_default_line_size(-1)
         self.assertEqual(xmlsec.base64_default_line_size(), size)
+
+
+class TestCallbacks(base.TestMemoryLeaks):
+    def setUp(self):
+        super().setUp()
+        xmlsec.cleanup_callbacks()
+
+    def _sign_doc(self):
+        root = self.load_xml("doc.xml")
+        sign = xmlsec.template.create(
+            root,
+            c14n_method=consts.TransformExclC14N,
+            sign_method=consts.TransformRsaSha1
+        )
+        xmlsec.template.add_reference(
+            sign, consts.TransformSha1, uri="cid:123456")
+
+        ctx = xmlsec.SignatureContext()
+        ctx.key = xmlsec.Key.from_file(
+            self.path("rsakey.pem"), format=consts.KeyDataFormatPem
+        )
+        ctx.sign(sign)
+        return sign
+
+    def _expect_sign_failure(self):
+        exc_info = pytest.raises(xmlsec.Error, self._sign_doc)
+        self.assertEqual(exc_info.value.args, (1, 'failed to sign'))
+
+    def _register_mismatch_callbacks(self, match_cb=lambda filename: False):
+        xmlsec.register_callbacks(
+            match_cb,
+            lambda filename: None,
+            lambda none, buf: 0,
+            lambda none: None,
+        )
+
+    def _register_match_callbacks(self):
+        xmlsec.register_callbacks(
+            lambda filename: filename == b'cid:123456',
+            lambda filename: BytesIO(b'<html><head/><body/></html>'),
+            lambda bio, buf: bio.readinto(buf),
+            lambda bio: bio.close(),
+        )
+
+    def _find(self, elem, *tags):
+        for tag in tags:
+            elem = elem.find(
+                '{{http://www.w3.org/2000/09/xmldsig#}}{}'.format(tag))
```

Comment:
I am no expert on the matter, but blame `lxml` here, never trusted its `ElementPath` anyway. XPath should work:
```suggestion
            elem = next(iter(elem.xpath("./xmldsig:{}".format(tag), namespaces={"xmldsig": "http://www.w3.org/2000/09/xmldsig#"})))
```

Reply:
This was indeed the issue in my local Fedora environment - no idea why it didn't show up on my Ubuntu derivative.